src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -317,13 +317,31 @@ public static int DivRem(int a, int b, out int result)
             // Restore to using % and / when the JIT is able to eliminate one of the idivs.
             // In the meantime, a * and - is measurably faster than an extra /.
 
+            // TODO: implement DivRem intrinsic for Mono
+#if !MONO
+            if (X86Base.IsSupported)
+            {
+                (int quotient, result) = X86Base.DivRem((uint)a, a >> 31, b);
```

Comment:
The signed divisions need to use `IDIV` instruction and can't use the unsigned `DIV`.

Reply:
Nvm, didn't notice that it will go to the correct overload based on the 2nd argument. But will the JIT be smart enough to use `CDQ/CDO` instructions here?
And constant divisors should not use the intrinsic at all.

Reply:
> And constant divisors should not use the intrinsic at all.

Do you have a godbolt example to show what should happen ideally when you need both quotient and remainder for constant divisors?

Reply:
For constant divisors, the current code below is the ideal (or at least pretty close):
```cs
int div = a / b; // optimized by RyuJIT to a multiplication (+potentially some shifts)
result = a - (div * b);
return div;

Reply:
So may be in JIT, we can check if the divisor is a constant, just do not detect it and proceed with fallback route.

Reply:
> And CSE also probably doesn't work with this?

Correct, it doesn't CSE because of the following:

https://github.com/dotnet/runtime/blob/2aebbf75167b3aa21cae23e6fd702688dd724663/src/coreclr/jit/valuenum.cpp#L9522-L9531

@jakobbotsch

Reply:
Have you checked the codegen for callsites that use constant divisors for `Math.DivRem`? I don't see any mitigation for such cases, so I would NOT change `Math.DivRem` to use the new HW intrinsics in this PR.

Reply:
```asm
G_M59029_IG02:
       mov      edx, ecx
       sar      edx, 31
       mov      r8d, 6
       mov      eax, ecx
       idiv     edx:eax, r8d
       mov      dword ptr [rsp], eax
       mov      dword ptr [rsp+04H], edx
       mov      rax, qword ptr [rsp]
```

Reply:
That's definitely worse and not what's expected by the users of the `Math.DivRem` API - so unless there's a good solution found for this, I'm still against using the intrinsic from `Math.DivRem`.